apps/cms.c

```diff
@@ -885,10 +892,31 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r',
-                          binary_files ? FORMAT_BINARY : informat);
-    if (in == NULL)
-        goto end;
+    if (digesthex != NULL) {
+        if (infile != NULL || !(flags & CMS_DETACHED)) {
+            BIO_printf(bio_err,
+                       "Cannot use -digest when -in or -nodetach is given\n");
+            goto end;
+        }
+        if (operation != SMIME_SIGN) {
+            BIO_printf(bio_err,
+                       "Warning: -digest is ignored for non-signing operation\n");
+        } else {
+            digestbin = OPENSSL_hexstr2buf(digesthex, &digestlen);
+            if (digestbin == NULL) {
+                BIO_printf(bio_err,
+                           "Invalid hex value after -digest\n");
+                goto end;
+            }
+        }
+    }
+
+    if (digestbin == NULL) {
```

Comment:
`digestbin == NULL` is not wrong here,
but I think `digesthex == NULL` would be more straightforward to understand.

Reply:
It's not enough though. If the "Warning: -digest is ignored for non-signing operation\n" is printed above, we have `digesthex != NULL` and `digestbin == NULL`.

I'll try to come up with a condition that is both correct and easy to understand.

Reply:
If we free digesthex and set it to NULL where the warning is printed, we can have `if (digesthex == NULL)` here. Do you like this idea?

Reply:
Here is an IMHO more straightforward solution:
Remove the warning part from the `if (digesthex != NULL)` block, 
return to the original `else` after the block,
and add 
```
    if (operation != SMIME_SIGN && digesthex != NULL) {
        BIO_printf(bio_err,
                   "Warning: -digest is ignored for non-signing operation\n");
        digesthex = NULL;
    }
```
before the `if (digesthex != NULL)`.